python autoinstaller is broken with python 3.12

mcatanzaro opened the pull request:
#### cb4faa2a3ee38601f41c1d60fcac002a83cec4ff
<pre>
python autoinstaller is broken with python 3.12
<a href="https://bugs.webkit.org/show_bug.cgi?id=260726">https://bugs.webkit.org/show_bug.cgi?id=260726</a>

Reviewed by Jonathan Bedard.

Currently git-webkit is broken with python 3.12 due to various problems
with dependencies that are already fixed in newer upstream versions. So,
update a few things.

Unfortunately, the newer library versions are incompatible with other
Python versions, so this means more conditional versions.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py:

Canonical link: <a href="https://commits.webkit.org/267415@main">https://commits.webkit.org/267415@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6df9e0c671848bfb0e043ac3de6ed247c102a069

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18301 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15494 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16710 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16985 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16717 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17127 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14288 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19079 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/16645 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14371 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14972 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21764 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15359 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15138 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19448 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15734 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13359 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/16403 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14934 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3954 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19303 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15559 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py:
```diff
@@ -53,7 +53,7 @@
 if sys.version_info > (3, 0):
     AutoInstall.register(Package('mock', Version(4)))
-    AutoInstall.register(Package('setuptools', Version(56, 0, 0)))
+    AutoInstall.register(Package('setuptools', Version(68, 1, 2)))
```
Based on EWS results, I think this will need to be more specifically differentiated based on Python  version.

mcatanzaro replied:
OK, I think you're right. I didn't think this was the problem at first because I saw the python 3 version of the tests was failing here, not the python 2 version. But in fact, the old version of setuptools requires python 3.6, while the new version requires 3.8. The EWS must be using python 3.6 or 3.7. That would explain it.

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py:
```diff
@@ -77,12 +77,12 @@
 
 AutoInstall.register(Package('pyparsing', Version(2, 4, 7)))
-AutoInstall.register(Package('requests', Version(2, 24)))
+AutoInstall.register(Package('requests', Version(2, 31, 0)))
```
Looks like this version doesn't support Python 2. That's fine in general, we just should let Python 2 use the old version still.

mcatanzaro replied:
Oops. At least that's easy to fix.

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py:
```diff
@@ -77,12 +82,22 @@
 AutoInstall.register(Package('pyparsing', Version(2, 4, 7)))
-AutoInstall.register(Package('requests', Version(2, 24)))
+
+if sys.version_info > (3, 7):
```
Can we maybe change this to `3.9`? We will eventually need to sort out the `urllib3` import problem, but we can kick that can down the road until we're forced to address it.

mcatanzaro replied:
We don't need to use newer versions of anything until 3.12, so I might as well just specify 3.12 instead of 3.9? What's special about version 3.9? (I had picked 3.7 simply because that is the minimum supported python version according to the pypi metadata for these libraries.)

mcatanzaro replied:
OK, changing this to 3.9 won't actually work, because that's the version the broken EWS is already using. :(

mcatanzaro replied:
I'm going to change these all to 3.12 for the benefit of unbreaking git-webkit so I can continue normal WebKit development.

It's a real shame we don't understand why this is not working with python 3.9.

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py:
```diff
@@ -77,12 +82,22 @@
 AutoInstall.register(Package('tblib', Version(1, 7, 0)))
-AutoInstall.register(Package('urllib3', Version(1, 25, 10)))
+
+if sys.version_info > (3, 7):
```
Ditto on the `3.9` comment.

mcatanzaro committed:
python autoinstaller is broken with python 3.12
https://bugs.webkit.org/show_bug.cgi?id=260726

Reviewed by Jonathan Bedard.

Currently git-webkit is broken with python 3.12 due to various problems
with dependencies that are already fixed in newer upstream versions. So,
update a few things.

Unfortunately, the newer library versions are incompatible with other
Python versions, so this means more conditional versions.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py:

Canonical link: https://commits.webkit.org/267415@main
